MagneticField/ParametrizedEngine/test/testParabolicParametrizedMagneticField.cc

```diff
@@ -0,0 +1,64 @@
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/FileInPath.h"
+#include "DataFormats/GeometryVector/interface/GlobalPoint.h"
+#include "DataFormats/GeometryVector/interface/GlobalVector.h"
+#include "DataFormats/FWLite/interface/Record.h"
+
+#include "MagneticField/ParametrizedEngine/interface/alpaka/ParabolicParametrizedMagneticField.h"
+
+#include <iostream>
+#include <fstream>
+#include <Eigen/Core>
+
+using namespace edm;
+using namespace std;
+using namespace ALPAKA_ACCELERATOR_NAMESPACE::MagneticFieldParabolicPortable;
+using Vector3f = Eigen::Matrix<float, 3, 1>;
+
+int main() {
+  ifstream file;
+  edm::FileInPath mydata("MagneticField/Engine/data/Regression/referenceField_160812_RII_3_8T.bin");
+  file.open(mydata.fullPath().c_str(), ios::binary);
+
+  float resolution = 0.0001;
+  float maxdelta = 0.;
+  int fail = 0;
+  int count = 0;
+
+  float px, py, pz;
+  float bx, by, bz;
+  GlobalPoint gp;
+
+  int numberOfPoints = 100;
+  do {
+    if (!(file.read((char*)&px, sizeof(float)) && file.read((char*)&py, sizeof(float)) &&
+          file.read((char*)&pz, sizeof(float)) && file.read((char*)&bx, sizeof(float)) &&
+          file.read((char*)&by, sizeof(float)) && file.read((char*)&bz, sizeof(float))))
+      break;
+    gp = GlobalPoint(px, py, pz);
+
+    if (gp.perp2() > Parameters::max_radius2 || gp.z() > Parameters::max_z)
+      continue;
+
+    GlobalVector referenceB(bx, by, bz);
+    GlobalVector deviceB(0, 0, MagneticFieldAtPoint(Vector3f(px, py, pz)));
+    if ((referenceB - deviceB).mag() > resolution) {
+      ++fail;
+      float delta = (referenceB - deviceB).mag();
+      if (delta > maxdelta)
+        maxdelta = delta;
+      if (fail < 10) {
+        cout << " Discrepancy at: # " << count + 1 << " " << gp << " R " << gp.perp() << " Phi " << gp.phi()
+             << " delta : " << referenceB - deviceB << " " << delta << endl;
+        cout << " Old: " << deviceB << " New: " << referenceB << endl;
+      } else if (fail == 10) {
+        cout << "..." << endl;
+      }
+    }
+    count++;
+  } while (count < numberOfPoints);
+
+  if (fail != 0)
+    //throw cms::Exception("RegressionFailure") << "MF regression found: " << fail << " failures";
+    cout << "MF regression found: " << fail << " failures";
```

Comment:
The test should return `EXIT_FAILURE` if `(fail != 0)`, so that scram can notice the failure.

Reply:
And you may need to add
```c++
#include <cstdlib>
```
to define that constant.

Reply:
> The test should return `EXIT_FAILURE` if `(fail != 0)`, so that scram can notice the failure.

Would it be enough to throw an Exception as in the commented line 62?
I momentarily commented that line because it makes compilation fail with:
```
>> Compiling  src/MagneticField/ParametrizedEngine/test/testParabolicParametrizedMagneticField.cc
src/MagneticField/ParametrizedEngine/test/testParabolicParametrizedMagneticField.cc: In function 'int main()':
src/MagneticField/ParametrizedEngine/test/testParabolicParametrizedMagneticField.cc:62:45: error: invalid use of incomplete type 'class cms::Exception'
   62 |     throw cms::Exception("RegressionFailure") << "MF regression found: " << fail << " failures";
      |                                             ^
In file included from src/MagneticField/ParametrizedEngine/test/testParabolicParametrizedMagneticField.cc:5:
/cvmfs/cms.cern.ch/el9_amd64_gcc12/cms/cmssw/CMSSW_14_1_0_pre0/src/DataFormats/FWLite/interface/Record.h:42:9: note: forward declaration of 'class cms::Exception'
   42 |   class Exception;
      |         ^~~~~~~~~
gmake: *** [config/SCRAM/GMake/Makefile.rules:1824: tmp/el9_amd64_gcc12/src/MagneticField/ParametrizedEngine/test/testParabolicParametrizedMagneticField/testParabolicParametrizedMagneticField.cc.o] Error 1

```

Reply:
An exception should also be OK.
Either way, I would suggest that you _make_ the test fail (for example comparing with a wrong result) and check how `scram b runtests` reports the failure.

Reply:
Sure. Any hint on how to solve [this compilation error](https://github.com/cms-sw/cmssw/pull/44360#discussion_r1665417448)?

Reply:
I suppose you need to add`
#include "FWCore/Utilities/interface/Exception.h"`

Reply:
With that `#include` it compiles but the Building fails with
```
>> Compiling  src/MagneticField/ParametrizedEngine/test/testParabolicParametrizedMagneticField.cc
>> Building binary testParabolicParametrizedMagneticField
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/../lib/gcc/x86_64-redhat-linux-gnu/12.3.1/../../../../x86_64-redhat-linux-gnu/bin/ld.bfd: tmp/el9_amd64_gcc12/src/MagneticField/ParametrizedEngine/test/testParabolicParametrizedMagneticField/cccPe3ai.ltrans0.ltrans.o: in function `std::ostream& operator<< <float>(std::ostream&, Basic3DVector<float> const&) [clone .isra.0]':
<artificial>:(.text+0xd): undefined reference to `geometryDetails::print3D(std::ostream&, double, double, double)'
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/../lib/gcc/x86_64-redhat-linux-gnu/12.3.1/../../../../x86_64-redhat-linux-gnu/bin/ld.bfd: tmp/el9_amd64_gcc12/src/MagneticField/ParametrizedEngine/test/testParabolicParametrizedMagneticField/cccPe3ai.ltrans0.ltrans.o: in function `main':
<artificial>:(.text.startup+0x66): undefined reference to `edm::FileInPath::FileInPath(char const*)'
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/../lib/gcc/x86_64-redhat-linux-gnu/12.3.1/../../../../x86_64-redhat-linux-gnu/bin/ld.bfd: <artificial>:(.text.startup+0x81): undefined reference to `edm::FileInPath::fullPath[abi:cxx11]() const'
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/../lib/gcc/x86_64-redhat-linux-gnu/12.3.1/../../../../x86_64-redhat-linux-gnu/bin/ld.bfd: <artificial>:(.text.startup+0x305): undefined reference to `edm::FileInPath::~FileInPath()'
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/../lib/gcc/x86_64-redhat-linux-gnu/12.3.1/../../../../x86_64-redhat-linux-gnu/bin/ld.bfd: <artificial>:(.text.startup+0x54e): undefined reference to `cms::Exception::Exception(char const*)'
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/../lib/gcc/x86_64-redhat-linux-gnu/12.3.1/../../../../x86_64-redhat-linux-gnu/bin/ld.bfd: <artificial>:(.text.startup+0x58c): undefined reference to `cms::Exception::Exception(cms::Exception const&)'
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/../lib/gcc/x86_64-redhat-linux-gnu/12.3.1/../../../../x86_64-redhat-linux-gnu/bin/ld.bfd: <artificial>:(.text.startup+0x596): undefined reference to `cms::Exception::~Exception()'
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/../lib/gcc/x86_64-redhat-linux-gnu/12.3.1/../../../../x86_64-redhat-linux-gnu/bin/ld.bfd: <artificial>:(.text.startup+0x59d): undefined reference to `cms::Exception::~Exception()'
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/../lib/gcc/x86_64-redhat-linux-gnu/12.3.1/../../../../x86_64-redhat-linux-gnu/bin/ld.bfd: <artificial>:(.text.startup+0x5a4): undefined reference to `typeinfo for cms::Exception'
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/../lib/gcc/x86_64-redhat-linux-gnu/12.3.1/../../../../x86_64-redhat-linux-gnu/bin/ld.bfd: <artificial>:(.text.startup+0x5c0): undefined reference to `edm::FileInPath::~FileInPath()'
/cvmfs/cms.cern.ch/el9_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/../lib/gcc/x86_64-redhat-linux-gnu/12.3.1/../../../../x86_64-redhat-linux-gnu/bin/ld.bfd: <artificial>:(.text.startup+0x5e4): undefined reference to `cms::Exception::~Exception()'
collect2: error: ld returned 1 exit status
>> Deleted: tmp/el9_amd64_gcc12/src/MagneticField/ParametrizedEngine/test/testParabolicParametrizedMagneticField/testParabolicParametrizedMagneticField
gmake: *** [config/SCRAM/GMake/Makefile.rules:1824: tmp/el9_amd64_gcc12/src/MagneticField/ParametrizedEngine/test/testParabolicParametrizedMagneticField/testParabolicParametrizedMagneticField] Error 1
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2

```

Reply:
You then also have to link the library, adding 

`<use name="FWCore/Utilities"/>`

in `test/BuildFile.xml`

Reply:
It worked, thanks!